Frameworks/CoreGraphics/CGGradient.mm

```diff
@@ -18,186 +18,170 @@
 #import <StubReturn.h>
 #import <Starboard.h>
 #import <CoreGraphics/CGContext.h>
+#import <CFCppBase.h>
+#import <vector>
 #import "CGColorSpaceInternal.h"
 #import "CGGradientInternal.h"
 #import "UIColorInternal.h"
-#import "_CGLifetimeBridgingType.h"
+#import <algorithm>
+#import <iterator>
 
-@interface CGNSGradient : _CGLifetimeBridgingType
-@end
-
-@implementation CGNSGradient
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wobjc-missing-super-calls"
-- (void)dealloc {
-    delete (__CGGradient*)self;
-}
-#pragma clang diagnostic pop
-@end
-
-__CGGradient::__CGGradient() : _components(NULL), _locations(NULL) {
-    object_setClass((id)this, [CGNSGradient class]);
-}
+struct __CGGradient : CoreFoundation::CppBase<__CGGradient> {
+    __CGGradient() : _colorSpace(nullptr) {
+    }
 
-__CGGradient::~__CGGradient() {
-    if (_components) {
-        delete[] _components;
-        _components = NULL;
+    __CGGradient(CGColorSpaceRef colorSpace) {
+        SetColorSpace(colorSpace);
     }
-    if (_locations) {
-        delete[] _locations;
-        _locations = NULL;
+
+    ~__CGGradient() {
     }
-}
 
-void __CGGradient::initWithColorComponents(const float* components, const float* locations, size_t count, CGColorSpaceRef colorspace) {
-    __CGColorSpace* cs = (__CGColorSpace*)colorspace;
-    DWORD componentCount = 0;
-
-    _colorSpaceModel = cs->colorSpaceModel;
-
-    switch (_colorSpaceModel) {
-        case kCGColorSpaceModelRGB:
-            componentCount = 4;
-            _format = _ColorABGR;
-            break;
-        case kCGColorSpaceModelMonochrome:
-            componentCount = 2;
-            _format = _ColorGrayscale;
-            break;
-        default:
-            UNIMPLEMENTED_WITH_MSG("Unsupported colorspace used to create gradiant.");
-            break;
+    inline void SetColorSpace(CGColorSpaceRef colorSpace) {
+        _colorSpace = colorSpace;
     }
 
-    _components = new float[count * componentCount];
-    memcpy(_components, components, sizeof(float) * count * componentCount);
+    inline CGColorSpaceRef GetColorSpace() {
+        return _colorSpace;
+    }
 
-    _locations = new float[count];
+    inline std::vector<CGFloat> GetColorComponents() {
+        return _colorComponents;
```

Comment:
Actually, I think the appropriate model for this API is:

```cpp
std::vector<CGFloat> values;
CFIndex count;
_CGGradientGetColorComponents(gradient, nullptr, &count);
values.resize(count);
_CGGradientGetColorComponents(gradient, values.data(), &count);
```

This would be just like the CFArray, CFDictionary, etc. APIs that return caller-unknown counts of members via out pointer.